// bitcoin-sfox-client/lib/api.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var assert = require('assert');
var Exchange = require('bitcoin-exchange-client');

var API = function (_Exchange$API) {
  _inherits(API, _Exchange$API);

  function API() {
    _classCallCheck(this, API);

    var _this = _possibleConstructorReturn(this, (API.__proto__ || Object.getPrototypeOf(API)).call(this));

    _this._apiKey = null;
    _this._partnerId = 'blockchain';
    _this._accountToken = null;
    return _this;
  }

  _createClass(API, [{
    key: '_url',
    value: function _url(subdomain, version, endpoint) {
      assert(endpoint, 'endpoint required');
      version = version || 'v2';
      subdomain = subdomain || 'api';
      return 'https://' + subdomain + (this._production ? '' : '.staging') + '.sfox.com/' + version + '/partner/' + this._partnerId + '/' + endpoint;
    }
  }, {
    key: 'GET',
    value: function GET(endpoint, data, version, subdomain) {
      return this._request('GET', endpoint, version, subdomain, data);
    }
  }, {
    key: 'authGET',
    value: function authGET(endpoint, data, version, subdomain) {
      return this._request('GET', endpoint, version, subdomain, data, true);
    }
  }, {
    key: 'POST',
    value: function POST(endpoint, data, version, subdomain) {
      return this._request('POST', endpoint, version, subdomain, data);
    }
  }, {
    key: 'authPOST',
    value: function authPOST(endpoint, data, version, subdomain) {
      return this._request('POST', endpoint, version, subdomain, data, true);
    }
  }, {
    key: 'PATCH',
    value: function PATCH(endpoint, data, version, subdomain) {
      return this._request('PATCH', endpoint, version, subdomain, data);
    }
  }, {
    key: 'authPATCH',
    value: function authPATCH(endpoint, data, version, subdomain) {
      return this._request('PATCH', endpoint, version, subdomain, data, true);
    }
  }, {
    key: '_request',
    value: function _request(method, endpoint, version, subdomain, data, authorized) {
      assert(this._apiKey, 'API key required');
      assert(!authorized || this.hasAccount, "Can't make authorized request without an account");

      var url = this._url(subdomain, version, endpoint);

      var headers = {};

      if (subdomain !== 'quotes') {
        headers['X-SFOX-PARTNER-ID'] = this._apiKey;
      }

      if (authorized) {
        headers['Authorization'] = 'Bearer ' + this._accountToken;
      }

      return _get(API.prototype.__proto__ || Object.getPrototypeOf(API.prototype), '_request', this).call(this, method, url, data, headers);
    }
  }, {
    key: 'apiKey',
    get: function get() {
      return this._apiKey;
    },
    set: function set(value) {
      this._apiKey = value;
    }
  }, {
    key: 'partnerId',
    get: function get() {
      return this._partnerId;
    },
    set: function set(value) {
      this._partnerId = value;
    }
  }, {
    key: 'accountToken',
    get: function get() {
      return this._accountToken;
    }
  }, {
    key: 'hasAccount',
    get: function get() {
      return Boolean(this.accountToken);
    }
  }]);

  return API;
}(Exchange.API);

module.exports = API;
