// bitcoin-exchange-client/lib/exchange.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var assert = require('assert');
var Helpers = require('./helpers');

// This is poor man's abstract class:
// https://gist.github.com/Zodiase/af44115098b20d69c531

var Exchange = function () {
  function Exchange(obj, delegate, api, TradeClass, QuoteClass, PaymentMediumClass) {
    _classCallCheck(this, Exchange);

    assert(this.constructor !== Exchange, 'Abstract Class');
    assert(delegate, 'ExchangeDelegate required');
    assert(api, 'API class required');
    assert(TradeClass, 'Trade class required');
    assert(QuoteClass, 'Quote class required');
    assert(PaymentMediumClass, 'PaymentMedium class required');
    assert(QuoteClass.getQuote, 'Quote.getQuote missing');
    this._delegate = delegate;
    this._api = api;
    this._trades = [];
    this._TradeClass = TradeClass;
    this._QuoteClass = QuoteClass;
    this._PaymentMediumClass = PaymentMediumClass;

    this._trades = [];

    if (obj.trades) {
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = obj.trades[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var tradeObj = _step.value;

          var trade = new TradeClass(tradeObj, this._api, delegate);
          trade._getQuote = QuoteClass.getQuote; // Prevents circular dependency
          trade.debug = this._debug;
          this._trades.push(trade);
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }
  }

  _createClass(Exchange, [{
    key: 'getBuyMethods',
    value: function getBuyMethods() {
      return this._PaymentMediumClass.getAll(undefined, 'BTC', this._api);
    }
  }, {
    key: 'getSellMethods',
    value: function getSellMethods() {
      return this._PaymentMediumClass.getAll('BTC', undefined, this._api);
    }
  }, {
    key: 'getBuyQuote',
    value: function getBuyQuote(amount, baseCurrency, quoteCurrency) {
      assert(baseCurrency, 'Specify base currency');
      assert(baseCurrency !== 'BTC' || quoteCurrency, 'Specify quote currency');
      // istanbul ignore else
      if (baseCurrency !== 'BTC') {
        quoteCurrency = 'BTC';
      }
      return this._QuoteClass.getQuote(this._api, this._delegate, -amount, baseCurrency, quoteCurrency, this._debug);
    }
  }, {
    key: 'getSellQuote',
    value: function getSellQuote(amount, baseCurrency, quoteCurrency) {
      return this.getBuyQuote(-amount, baseCurrency, quoteCurrency);
    }
  }, {
    key: 'getTrades',
    value: function getTrades(QuoteClass) {
      var _this = this;

      assert(QuoteClass, 'QuoteClass required');
      var save = function save() {
        return _this.delegate.save.bind(_this.delegate)().then(function () {
          return _this._trades;
        });
      };
      var update = function update(tradeObjects) {
        var _iteratorNormalCompletion2 = true;
        var _didIteratorError2 = false;
        var _iteratorError2 = undefined;

        try {
          var _loop = function _loop() {
            var tradeObj = _step2.value;

            var id = _this._TradeClass.idFromAPI(tradeObj);

            var trade = _this._trades.find(function (trade) {
              return String(trade.id).toLowerCase() === String(id).toLowerCase();
            });

            if (!trade) {
              // We don't cache e.g. cancelled trades
              trade = new _this._TradeClass(null, _this._api, _this.delegate);
              trade._getQuote = QuoteClass.getQuote; // Prevents circular dependency
              _this._trades.push(trade);
            }

            trade.setFromAPI(tradeObj);
          };

          for (var _iterator2 = tradeObjects[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
            _loop();
          }
        } catch (err) {
          _didIteratorError2 = true;
          _iteratorError2 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion2 && _iterator2.return) {
              _iterator2.return();
            }
          } finally {
            if (_didIteratorError2) {
              throw _iteratorError2;
            }
          }
        }
      };
      var process = function process() {
        var _iteratorNormalCompletion3 = true;
        var _didIteratorError3 = false;
        var _iteratorError3 = undefined;

        try {
          for (var _iterator3 = _this._trades[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
            var _trade = _step3.value;

            _trade.process(_this._trades);
          }
        } catch (err) {
          _didIteratorError3 = true;
          _iteratorError3 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion3 && _iterator3.return) {
              _iterator3.return();
            }
          } finally {
            if (_didIteratorError3) {
              throw _iteratorError3;
            }
          }
        }
      };
      return this._TradeClass.fetchAll(this._api).then(update).then(process).then(save);
    }
  }, {
    key: 'monitorPayments',
    value: function monitorPayments() {
      this._TradeClass.monitorPayments(this._trades, this.delegate);
    }
  }, {
    key: 'api',
    get: function get() {
      return this._api;
    }
  }, {
    key: 'debug',
    get: function get() {
      return this._debug;
    },
    set: function set(value) {
      this._debug = Boolean(value);
      this._delegate.debug = Boolean(value);
      var _iteratorNormalCompletion4 = true;
      var _didIteratorError4 = false;
      var _iteratorError4 = undefined;

      try {
        for (var _iterator4 = this.trades[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
          var _trade2 = _step4.value;

          _trade2.debug = Boolean(value);
        }
      } catch (err) {
        _didIteratorError4 = true;
        _iteratorError4 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion4 && _iterator4.return) {
            _iterator4.return();
          }
        } finally {
          if (_didIteratorError4) {
            throw _iteratorError4;
          }
        }
      }
    }
  }, {
    key: 'user',
    get: function get() {
      return this._user;
    }
  }, {
    key: 'autoLogin',
    get: function get() {
      return this._auto_login;
    },
    set: function set(value) {
      assert(Helpers.isBoolean(value), 'Boolean');
      this._auto_login = value;
      this.delegate.save.bind(this.delegate)();
    }
  }, {
    key: 'trades',
    get: function get() {
      return this._trades;
    }
  }, {
    key: 'delegate',
    get: function get() {
      return this._delegate;
    }
  }, {
    key: 'profile',
    get: function get() {
      return this._profile;
    }
  }]);

  return Exchange;
}();

module.exports = Exchange;
