// bitcoin-coinify-client/lib/api.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var assert = require('assert');
var Exchange = require('bitcoin-exchange-client');

var API = function (_Exchange$API) {
  _inherits(API, _Exchange$API);

  function API() {
    _classCallCheck(this, API);

    var _this = _possibleConstructorReturn(this, (API.__proto__ || Object.getPrototypeOf(API)).call(this));

    _this._offlineToken = null;
    _this._loginExpiresAt = null;
    return _this;
  }

  _createClass(API, [{
    key: 'login',
    value: function login() {
      var self = this;

      var promise = new Promise(function (resolve, reject) {
        assert(self._offlineToken, 'Offline token required');

        var loginSuccess = function loginSuccess(res) {
          self._access_token = res.access_token;
          self._loginExpiresAt = new Date(new Date().getTime() + res.expires_in * 1000);
          resolve();
        };

        var loginFailed = function loginFailed(e) {
          reject(e);
        };
        self.POST('auth', {
          grant_type: 'offline_token',
          offline_token: self._offlineToken
        }).then(loginSuccess).catch(loginFailed);
      });

      return promise;
    }
  }, {
    key: '_url',
    value: function _url(endpoint) {
      endpoint = endpoint || '';
      return 'https://app-api' + (this._sandbox ? '.sandbox' : '') + '.coinify.com/' + endpoint;
    }
  }, {
    key: '_request',
    value: function _request(method, endpoint, data, authorized) {
      assert(!authorized || this.isLoggedIn, "Can't make authorized request if not logged in");

      var headers = {};

      if (authorized) {
        headers['Authorization'] = 'Bearer ' + this._access_token;
      }

      return _get(API.prototype.__proto__ || Object.getPrototypeOf(API.prototype), '_request', this).call(this, method, this._url(endpoint), data, headers);
    }
  }, {
    key: '_authRequest',
    value: function _authRequest(method, endpoint, data) {
      var doRequest = function doRequest() {
        return this._request(method, endpoint, data, true);
      };

      if (this.isLoggedIn) {
        return doRequest.bind(this)();
      } else {
        return this.login().then(doRequest.bind(this));
      }
    }
  }, {
    key: 'GET',
    value: function GET(endpoint, data) {
      return this._request('GET', endpoint, data);
    }
  }, {
    key: 'authGET',
    value: function authGET(endpoint, data) {
      return this._authRequest('GET', endpoint, data);
    }
  }, {
    key: 'POST',
    value: function POST(endpoint, data) {
      return this._request('POST', endpoint, data);
    }
  }, {
    key: 'authPOST',
    value: function authPOST(endpoint, data) {
      return this._authRequest('POST', endpoint, data);
    }
  }, {
    key: 'PATCH',
    value: function PATCH(endpoint, data) {
      return this._request('PATCH', endpoint, data);
    }
  }, {
    key: 'authPATCH',
    value: function authPATCH(endpoint, data) {
      return this._authRequest('PATCH', endpoint, data);
    }
  }, {
    key: 'DELETE',
    value: function DELETE(endpoint, data) {
      return this._authRequest('DELETE', endpoint, data);
    }
  }, {
    key: 'isLoggedIn',
    get: function get() {
      // Debug: + 60 * 19 * 1000 expires the login after 1 minute
      var tenSecondsFromNow = new Date(new Date().getTime() + 10000);
      return Boolean(this._access_token) && this._loginExpiresAt > tenSecondsFromNow;
    }
  }, {
    key: 'offlineToken',
    get: function get() {
      return this._offlineToken;
    }
  }, {
    key: 'hasAccount',
    get: function get() {
      return Boolean(this.offlineToken);
    }
  }]);

  return API;
}(Exchange.API);

module.exports = API;
