// bitcoin-coinify-client/lib/quote.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PaymentMethod = require('./payment-medium');
var Exchange = require('bitcoin-exchange-client');
var Trade = require('./trade');

var Quote = function (_Exchange$Quote) {
  _inherits(Quote, _Exchange$Quote);

  function Quote(obj, api, delegate, debug) {
    _classCallCheck(this, Quote);

    var _this = _possibleConstructorReturn(this, (Quote.__proto__ || Object.getPrototypeOf(Quote)).call(this, api, delegate, Trade, PaymentMethod, debug));

    var expiresAt = new Date(obj.expiryTime);
    var timeOfRequest = new Date(obj.issueTime);

    // Debug, make quote expire in 15 seconds:
    // expiresAt = new Date(new Date().getTime() + 15 * 1000);

    _this._id = obj.id;
    _this._baseCurrency = obj.baseCurrency;
    _this._quoteCurrency = obj.quoteCurrency;
    _this._expiresAt = expiresAt;
    _this._timeOfRequest = timeOfRequest;

    if (_this._baseCurrency === 'BTC') {
      _this._baseAmount = Math.round(obj.baseAmount * 100000000);
      _this._quoteAmount = Math.round(obj.quoteAmount * 100);
    } else {
      _this._baseAmount = Math.round(obj.baseAmount * 100);
      _this._quoteAmount = Math.round(obj.quoteAmount * 100000000);
    }

    obj.baseAmount;
    return _this;
  }

  _createClass(Quote, [{
    key: 'expire',


    // QA tool
    value: function expire() {
      this._expiresAt = new Date(new Date().getTime() + 3 * 1000);
    }
  }, {
    key: 'paymentMediums',
    get: function get() {
      return this._paymentMediums;
    }
  }], [{
    key: 'getQuote',
    value: function getQuote(api, delegate, amount, baseCurrency, quoteCurrency, debug) {
      var processQuote = function processQuote(quote) {
        return new Quote(quote, api, delegate);
      };

      var getQuote = function getQuote(_baseAmount) {
        var getAnonymousQuote = function getAnonymousQuote() {
          return api.POST('trades/quote', {
            baseCurrency: baseCurrency,
            quoteCurrency: quoteCurrency,
            baseAmount: parseFloat(_baseAmount)
          });
        };

        var getQuote = function getQuote() {
          return api.authPOST('trades/quote', {
            baseCurrency: baseCurrency,
            quoteCurrency: quoteCurrency,
            baseAmount: parseFloat(_baseAmount)
          });
        };

        if (!api.hasAccount) {
          return getAnonymousQuote().then(processQuote);
        } else {
          return getQuote().then(processQuote);
        }
      };

      return _get(Quote.__proto__ || Object.getPrototypeOf(Quote), 'getQuote', this).call(this, amount, baseCurrency, quoteCurrency, ['BTC', 'EUR', 'GBP', 'USD', 'DKK'], debug).then(getQuote);
    }
  }]);

  return Quote;
}(Exchange.Quote);

module.exports = Quote;
